src/hotspot/share/opto/divnode.cpp

```diff
@@ -1203,44 +1203,78 @@ Node *ModINode::Ideal(PhaseGVN *phase, bool can_reshape) {
 }
 
 //------------------------------Value------------------------------------------
-const Type* ModINode::Value(PhaseGVN* phase) const {
+static const Type* mod_value(const PhaseGVN* phase, const Node* in1, const Node* in2, const BasicType bt, const Type* bottom) {
+  assert(bt == T_INT || bt == T_LONG, "unexpected basic type");
   // Either input is TOP ==> the result is TOP
-  const Type *t1 = phase->type( in(1) );
-  const Type *t2 = phase->type( in(2) );
-  if( t1 == Type::TOP ) return Type::TOP;
-  if( t2 == Type::TOP ) return Type::TOP;
+  const Type* t1 = phase->type(in1);
+  const Type* t2 = phase->type(in2);
+  if (t1 == Type::TOP) return Type::TOP;
+  if (t2 == Type::TOP) return Type::TOP;
 
   // We always generate the dynamic check for 0.
   // 0 MOD X is 0
-  if( t1 == TypeInt::ZERO ) return TypeInt::ZERO;
+  if (t1 == TypeInteger::zero(bt)) return t1;
+
   // X MOD X is 0
-  if (in(1) == in(2)) {
-    return TypeInt::ZERO;
+  if (in1 == in2) {
+    return TypeInteger::zero(bt);
   }
 
-  // Either input is BOTTOM ==> the result is the local BOTTOM
-  const Type *bot = bottom_type();
-  if( (t1 == bot) || (t2 == bot) ||
-      (t1 == Type::BOTTOM) || (t2 == Type::BOTTOM) )
-    return bot;
-
-  const TypeInt *i1 = t1->is_int();
-  const TypeInt *i2 = t2->is_int();
-  if( !i1->is_con() || !i2->is_con() ) {
-    if( i1->_lo >= 0 && i2->_lo >= 0 )
-      return TypeInt::POS;
-    // If both numbers are not constants, we know little.
-    return TypeInt::INT;
-  }
   // Mod by zero?  Throw exception at runtime!
-  if( !i2->get_con() ) return TypeInt::POS;
+  if (t2 == TypeInteger::zero(bt)) {
+    return TypeInt::TOP;
+  }
 
-  // We must be modulo'ing 2 float constants.
-  // Check for min_jint % '-1', result is defined to be '0'.
-  if( i1->get_con() == min_jint && i2->get_con() == -1 )
-    return TypeInt::ZERO;
+  const TypeInteger* i1 = t1->isa_integer(bt);
+  const TypeInteger* i2 = t2->isa_integer(bt);
+  if (i1 == nullptr || i2 == nullptr) {
+    return bottom;
+  }
+  if (i1->is_con() && i2->is_con()) {
+    // We must be modulo'ing 2 int constants.
+    // Check for min_jlong % '-1', result is defined to be '0'
+    // We don't need to check for min_jint % '-1' as its result is defined when using jlong.
+    if (i1->get_con_as_long(bt) == min_jlong && i2->get_con_as_long(bt) == -1) {
+      return TypeInteger::zero(bt);
+    }
+    return TypeInteger::make(i1->get_con_as_long(bt) % i2->get_con_as_long(bt), bt);
+  }
+  // The magnitude of the divisor is in range [1, 2^31] or [1, 2^63], depending on the BasicType.
+  // We know it isn't 0 as we handled that above.
+  // That means at least one value is nonzero, so its absolute value is bigger than zero.
```

Comment:
I'm actually struggling to follow this here. Can you define "magnitude" for the reader? Maybe there is some JVMS definition you can mention. And which "value" are you refering to, that is nonzero here?

Reply:
```suggestion
  // We checked that t2 is not the zero constant. Hence at least i2->_lo or i2->_hi must be non-zero,
  // and hence its its absoute value is bigger than zero. Hence, the magnitude of the divisor (i.e. the
  // largest absolute value for any value in i2) must be in the  range [1, 2^31] or [1, 2^63], depending
  // on the BasicType.
```